### PR TITLE
Initial PR for exchange discovery.

### PR DESCRIPTION
**Pull request by mavarley**
To address Issue #259 .

***
<a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/vc-api/pull/268.html" title="Last updated on Aug 17, 2022, 6:39 PM UTC (f66a184)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/vc-api/268/95cabd6...f66a184.html" title="Last updated on Aug 17, 2022, 6:39 PM UTC (f66a184)">Diff</a>

### CODE REVIEW

File: holder.yml [hunk 1]
```diff
@@ -56,6 +56,39 @@ paths:
           description: invalid input!
         "500":
           description: error!
+  /exchanges/:
+    get:
+      summary: provides a discovery endpoint for the exchanges supported by this sserver endpoint.
```
Comment (TallTed) [inline holder.yml:61]:
```suggestion
      summary: provides a discovery endpoint for the exchanges supported by this server endpoint.
```

File: holder.yml [hunk 2]
```diff
@@ -56,6 +56,39 @@ paths:
...
+      operationId: discoverExchanges
+      description: 
+        This endpoint returns a map of the exchange-ids (endpoints) supported by this server, 
+        and the associated protocol supported by that exchange endpoint in an 
+        "exchange-id" -> "protocol" map. The keys in the returned object MUST be a valid 
+        ExchangeId path parameter, and the value field MUST be a string that 
+        references the supported protocol on that endpoint. Protocol strings MUST be 
+        the protocol specification name or type, or a url to a spec. Clients consuming this
```
Comment (TallTed) [inline holder.yml:69]:
```suggestion
        the protocol specification name or type, or a URL to a spec. Clients consuming this 
```

File: holder.yml [hunk 3]
```diff
@@ -56,6 +56,39 @@ paths:
...
+          description: a map of the exchange-id endpoints to protocols those exchanges support.
+          content:
+            application/json:
+              schema:
+                type: object 
+                additionalProperties: 
+                  type: string
+              example:
+                {
+                  "credential-refresh" : "CredentialRefresh2020",
```
Comment (msporny) [inline holder.yml:83]: The way this is expressed in a VC is the following:
```
    "refreshService": {
      "type": "UnmediatedRefresh2021",
      "url": "https://example.edu/exchanges/refresh-degree",
      "validAfter": "2021-09-01T19:23:24Z"
    },
```

could map to this like so:

```
    "https://example.edu/exchanges/refresh-degree": {
      "type": "UnmediatedRefresh2021",
    },
```

Could be a proper JSON-LD object with semantics (not saying this is a good idea, just trying it out):

```
  "@context": ["https://w3id.org/vc/api/v1",  "https://w3id.org/vc-refresh-2021/v1"],
  "service": [{
    "type": "UnmediatedRefresh2021", 
    "serviceEndpoint": "https://example.edu/exchanges/refresh-degree"
  }]
```

**Review by OR13 — APPROVED** [review]
approved if change suggestions are accepted.

File: holder.yml [hunk 4]
```diff
@@ -56,6 +56,39 @@ paths:
           description: invalid input!
         "500":
           description: error!
+  /exchanges/:
```
Comment (OR13) [inline holder.yml:59]:
```suggestion
  /exchanges:
```

**Review by OR13 — CHANGES_REQUESTED** [review]
I would like the response example to reflect real world examples, support paging, etc...

File: holder.yml [hunk 5]
```diff
@@ -56,6 +56,39 @@ paths:
...
+        in the value field. Clients are not expected to dynamically process the protocol specified.    
+      responses:
+        "200":
+          description: a map of the exchange-id endpoints to protocols those exchanges support.
+          content:
+            application/json:
+              schema:
+                type: object 
+                additionalProperties: 
+                  type: string
```
Comment (OR13) [inline holder.yml:80]: don't return a map... return an array with the potential to support paging, etc.

File: holder.yml [hunk 6]
```diff
@@ -56,6 +56,39 @@ paths:
...
+        list wishing to use an exchange endpoint MUST recognize and support the protocol identified
+        in the value field. Clients are not expected to dynamically process the protocol specified.    
+      responses:
+        "200":
+          description: a map of the exchange-id endpoints to protocols those exchanges support.
+          content:
+            application/json:
+              schema:
+                type: object 
+                additionalProperties:
```
Comment (OR13) [inline holder.yml:79]: not sure this is valid yaml definition... looks a bit odd.

File: holder.yml [hunk 7]
```diff
@@ -56,6 +56,39 @@ paths:
...
+        and the associated protocol supported by that exchange endpoint in an 
+        "exchange-id" -> "protocol" map. The keys in the returned object MUST be a valid 
+        ExchangeId path parameter, and the value field MUST be a string that 
+        references the supported protocol on that endpoint. Protocol strings MUST be 
+        the protocol specification name or type, or a URL to a spec. Clients consuming this 
+        list wishing to use an exchange endpoint MUST recognize and support the protocol identified
+        in the value field. Clients are not expected to dynamically process the protocol specified.    
+      responses:
+        "200":
+          description: a map of the exchange-id endpoints to protocols those exchanges support.
```
Comment (OR13) [inline holder.yml:74]:
```json
{
    "totalItems": 8,
    "tutorials": [...],
    "totalPages": 3,
    "currentPage": 1
}
```

or similar.

Comment (mavarley) [reply holder.yml:74]: what are "tutorials" ?

**Comment by msporny** [comment]
This PR is awaiting a response on pagination. I expect the same applies to GET in PR #271.

**Review by OR13 — CHANGES_REQUESTED** [review]
API still does not define `items` or consider a world where the response might be too large for one page... we don't need to define all the pagination machinery to approve the PR, but we should be able to see how it could be defined in the future.

**Review by OR13 — APPROVED** [review]

File: holder.yml [hunk 8]
```diff
@@ -56,6 +56,102 @@ paths:
...
+                        type: string
+                        description: 
+                          The index position for the next set of results (ie, index of the end of this list). 
+                          Examples could be a numerica value, a URL, or a value meaningful to the server.
+
+              example:
+              {
+                  "count": 3,
+                  "total": 8,
+                  "excahnges" : [
```
Comment (msporny) [inline holder.yml:133]:
```suggestion
                  "exchanges" : [
```

**Comment by msporny** [comment]
The group discussed this on the 2022-05-17 call. There were no objections to merging the PR.

**Comment by jandrieu** [comment]
How does this PR support the naming of the components for the endpoints?

This continues to be a failing. Without identifying which component MUST implement or MAY implement any particular endpoint, it is hard to understand what is happening.

**Comment by msporny** [comment]
Still waiting on a re-base on this @mavarley to resolve conflicts -- ping.

**Comment by msporny** [comment]
> Still waiting on a re-base on this @mavarley to resolve conflicts -- ping.

Second ping...

**Comment by msporny** [comment]
We discussed this on the 2022-08-16 call and it was decided that we will merge this PR by the following weekend. Either @mavarley will rebase and merge or @msporny will rebase and merge -- whoever gets to it first.

File: index.html [hunk 9]
```diff
@@ -654,7 +665,11 @@ <h4>Exchange Examples</h4>
 step 2 above.
           </li>
         </ol>
-
+        <p>
+The Holder App MAY call the App's exchange discovery endpoint to determine if
+the Holder App supports the App's protocol requriements on a particular endpoint,
```
Comment (TallTed) [inline index.html:670]:
```suggestion
the Holder App supports the App's protocol requirements on a particular endpoint, 
```

**Comment by mavarley** [comment]
I think all the merging and edit suggestion have been applied properly. Merging.

**Pull request merged**